Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs

```diff
@@ -92,7 +121,105 @@ public void Dispose()
             RetryPolicy retryPolicy,
             CancellationToken cancellationToken = default)
         {
-            return await this.sessionPool.Execute<T>(func, retryPolicy, cancellationToken);
+            if (this.isClosed)
+            {
+                this.logger.LogError(ExceptionMessages.DriverClosed);
+                throw new QldbDriverException(ExceptionMessages.DriverClosed);
+            }
+
+            bool replaceDeadSession = false;
+            int retryAttempt = 0;
+            while (true)
+            {
+                AsyncQldbSession session = null;
+                try
+                {
+                    if (replaceDeadSession)
+                    {
+                        session = await this.StartNewSession();
+                    }
+                    else
+                    {
+                        session = await this.GetSession();
+                    }
+
+                    T returnedValue = await session.Execute(func);
+                    this.ReleaseSession(session);
+                    return returnedValue;
+                }
+                catch (RetriableException re)
+                {
+                    // If initial session is invalid, always retry once with a new session.
+                    if (re.InnerException is InvalidSessionException && retryAttempt == 0)
+                    {
+                        this.logger.LogDebug("Initial session received from pool invalid. Retrying...");
+                        replaceDeadSession = true;
+                        retryAttempt++;
```

Comment:
We always increment the retryCount regardless, why not do this as the first step in the catch and not have it in two separate areas?

Reply:
Done